### ISSUE TITLE
DNNs for feature classification

### ISSUE THREAD

**Issue by mortonjt** [qiime2/q2-feature-classifier#129]
**Improvement Description**
Crazy idea.

What if we implemented a deep convolutional neural network for taxonomy assignment?

**Comments**
A couple of immediate advantages come up to mind

1. DNNs are likely to be much more accurate than random forests or naive bayes. There is loads of literature on this topic, and this is starting to become more popular in the context of sequence classification (see [here](https://www.biorxiv.org/node/105928.abstract) just for an example).
2. Using something like Tensorflow or Pytorch would yield immediate gains in speed up, since they are packaged with multi-threading / GPU support.  I'm writing this as I'm approaching 2 hours for running taxonomy assignment for 30k sequences.  There also are other advantages worth considering such as mini-batching, model diagnostics, ...

**References**
[here](https://www.biorxiv.org/node/105928.abstract)

**Comment by nbokulich** [comment]
I like it; this has been on the back of my mind for a while. Sounds like you already have a DNN taxonomy classifier prototype... if your experiment works might I propose we benchmark in tax-credit and go from there?

**Comment by gregcaporaso** [comment]
@mortonjt, still interested in this? I like the plan - I suspect we just need a developer.

**Comment by mortonjt** [comment]
Hi, we have the software, but limited bandwidth to pursue this. Happy to
provide guidance for next steps

**Comment by BenKaehler** [comment]
Hi both, we already tried this:

https://www.frontiersin.org/articles/10.3389/fmicb.2021.644487

In summary we found that NNs and some other approaches did no better than NB, and even found some reasons for it.

You may be able to do better, but I would hope that our work would mean that you're not starting from scratch.

**Comment by mortonjt** [comment]
Thanks for the reference.

Here is the paper that I was referring to :
https://www.nature.com/articles/s41587-023-01917-2

I'd be surprised that NNs didn't beat NB -- probably because it was
supervised rather than self-supervised.
Throwing language models at this problem will probably improve the accuracy
(albeit it does raise the question if taxonomic inference rather than
phylogenetic inference is more biologically relevant).

https://www.biorxiv.org/content/10.1101/2021.07.09.451778v1.full
https://www.sciencedirect.com/science/article/abs/pii/S0957417422012039
https://www.pnas.org/doi/pdf/10.1073/pnas.2122636119
https://academic.oup.com/sysbio/article/72/1/17/6575921

**Comment by gregcaporaso** [comment]
Thanks everyone. I'm thinking I'll close this one out since it sounds like we don't have plans to do this now. We can move any ongoing conversation to the forum. (Feel free to reopen if someone does want to work on moving this forward in this plugin.)
